pkg/util/lsof/lsof_linux.go

```diff
@@ -0,0 +1,383 @@
+	ofl := &openFilesLister{
+		pid: pid,
+
+		readlink: os.Readlink,
+		stat:     os.Stat,
+		lstat:    os.Lstat,
+	}
+
+	ofl.procPath = procPath()
+
+	fs, err := procfs.NewFS(ofl.procPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ofl.proc, err = fs.Proc(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	ofl.socketInfo = readSocketInfo(ofl.procPIDPath())
+
+	return ofl.openFiles()
+}
+
+func (ofl *openFilesLister) procPIDPath() string {
+	return fmt.Sprintf("%s/%d", ofl.procPath, ofl.pid)
+}
+
+func (ofl *openFilesLister) openFiles() (Files, error) {
+	var files Files
+
+	// open files, socket, pipe (everything with a file descriptor, from /proc/<pid>/fd)
+	openFDFiles, err := ofl.fdMetadata()
+	if err != nil {
+		log.Debugf("Failed to get open FDs for pid %d: %s", ofl.pid, err)
+	} else {
+		files = append(files, openFDFiles...)
+	}
+
+	// memory mapped files, code, regions (from /proc/<pid>/maps)
+	mmapFiles, err := ofl.mmapMetadata()
+	if err != nil {
+		log.Debugf("Failed to get memory maps for pid %d: %s", ofl.pid, err)
+	} else {
+		files = append(files, mmapFiles...)
+	}
+
+	return files, nil
+}
+
+func (ofl *openFilesLister) mmapMetadata() (Files, error) {
+	maps, err := ofl.proc.ProcMaps()
+	if err != nil {
+		return nil, err
+	}
+
+	cwd, _ := ofl.proc.Cwd()
+
+	var files Files
+	for i, m := range maps {
+		if i > 0 && m.Pathname == maps[i-1].Pathname {
+			// skip duplicate entries
+			continue
+		}
+		if m.Pathname == "" {
+			// anonymous mapping
+			continue
+		}
+
+		if m.Dev == 0 || m.Inode == 0 {
+			// virtual memory region, eg. [heap], [stack], [vvar], etc
+			continue
+		}
+
+		file := File{
+			Name:     m.Pathname,
+			OpenPerm: permToString(m.Perms),
+		}
+
+		if file.Type, file.FilePerm, file.Size, _ = fileStats(ofl.stat, m.Pathname); file.Type == "" {
+			continue
+		}
+		file.Fd = mmapFD(m.Pathname, file.Type, cwd)
+
+		files = append(files, file)
+	}
+
+	return files, nil
+}
+
+func permToString(perms *procfs.ProcMapPermissions) string {
+	s := ""
+
+	for _, perm := range []struct {
+		set       bool
+		charSet   string
+		charUnset string
+	}{
+		{perms.Read, "r", "-"},
+		{perms.Write, "w", "-"},
+		{perms.Execute, "x", "-"},
+		{perms.Private, "p", ""},
+		{perms.Shared, "s", ""},
+	} {
+		if perm.set {
+			s += perm.charSet
+		} else {
+			s += perm.charUnset
+		}
+	}
+
+	return s
+}
+
+func mmapFD(path string, ty, cwd string) string {
+	/*
+		cwd  current working directory;
+		ltx  shared library text (code and data);
+		mem  memory-mapped file;
+		mmap memory-mapped device;
+		rtd  root directory;
+		txt  program text (code and data);
+	*/
+	fd := "unknown"
+	if ty == "REG" {
+		// knowing whether the file is memory mapped or program text would require knowing permissions
+		// which aren't parsed by gopsutil, so we just assume it's memory mapped
+		fd = "mem"
+	} else if ty == "DIR" {
+		if path == "/" {
+			fd = "rtd"
+		} else if cwd != "" && path == cwd {
+			fd = "cwd"
+		}
+	}
+	return fd
```

Comment:
I might find this version: 
```go
fd := "unknown"

if ty == "REG" {
		// knowing whether the file is memory mapped or program text would require knowing permissions
		// which aren't parsed by gopsutil, so we just assume it's memory mapped
		fd = "mem"
		return fd
}

if ty == "DIR" {
		if path == "/" {
			fd = "rtd"
		} else if cwd != "" && path == cwd {
			fd = "cwd"
		}
		return fd
}

return fd
```

A bit readable. Would you consider changing to that version?

Reply:
That's much better 👍